src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl

```diff
@@ -1,40 +1,240 @@
+    const long W_2 = convert_long(INPUT0_SIZE_X) * 2l;
+    y_d = (y_d % H_2 + H_2) % H_2;
+    x_d = (x_d % W_2 + W_2) % W_2;
+    const size_t y = (size_t)(y_d >= H ? H_2 - 1 - y_d : y_d);
+    const size_t x = (size_t)(x_d >= W ? W_2 - 1 - x_d : x_d);
+    return get_data_single_value(data, n, c, y, x);
+}
+#undef reflection_data_no_align
 #endif
-#ifdef INTERPOLATION_MODE_NEAREST
-    printf("INTERPOLATION_MODE_NEAREST = true\n");
+#else
+#error [clDNN grid_sample_ref.cl]: undefined padding mode
 #endif
 
-#ifdef PADDING_MODE_ZEROS
-    printf("PADDING_MODE_ZEROS = true\n");
+#define get_padded FUNC_CALL(get_padded)
+
+#if defined(INTERPOLATION_MODE_BILINEAR)
+#define bilinear FUNC(interpolate)
+inline data_et bilinear(const data_t* data, const size_t n, const size_t c, const grid_et y_n, const grid_et x_n) {
+    const grid_et y_d = denormalize(y_n, INPUT0_SIZE_Y);
+    const grid_et x_d = denormalize(x_n, INPUT0_SIZE_X);
+    const grid_et y_topleft = floor(y_d);
+    const grid_et x_topleft = floor(x_d);
+    const grid_et dy = y_d - y_topleft;
+    const grid_et dx = x_d - x_topleft;
+    const data_et v00 = get_padded(data, n, c, y_topleft, x_topleft);
+    const data_et v01 = get_padded(data, n, c, y_topleft, x_topleft + 1);
+    const data_et v10 = get_padded(data, n, c, y_topleft + 1, x_topleft);
+    const data_et v11 = get_padded(data, n, c, y_topleft + 1, x_topleft + 1);
+
+    const data_et q0 = (1 - dx) * v00 + dx * v01;
+    const data_et q1 = (1 - dx) * v10 + dx * v11;
+    return dy * q1 + (1 - dy) * q0;
+}
+#undef bilinear
+
+#elif defined(INTERPOLATION_MODE_NEAREST)
+#define nearest FUNC(interpolate)
+inline data_et nearest(const data_t* data, const size_t n, const size_t c, const grid_et y_n, const grid_et x_n) {
+    const grid_et y_nearest = rint(denormalize(y_n, INPUT0_SIZE_Y));
+    const grid_et x_nearest = rint(denormalize(x_n, INPUT0_SIZE_X));
+    return get_padded(data, n, c, y_nearest, x_nearest);
+}
+#undef nearest
+
+#elif defined(INTERPOLATION_MODE_BICUBIC)
+
+typedef MAKE_VECTOR_TYPE(INPUT1_TYPE, 4) vector_grid_4_t;
+typedef MAKE_VECTOR_TYPE(INPUT0_TYPE, 4) vector_data_4_t;
+typedef MAKE_VECTOR_TYPE(INPUT0_TYPE, 16) matrix_data_4x4_t;
+
+inline vector_grid_4_t FUNC(cubic_coeffs)(const data_et r, const data_et A) {
+    vector_grid_4_t v;
+    v[0] = ((A * (r + 1) - 5 * A) * (r + 1) + 8 * A) * (r + 1) - 4 * A;
+    v[1] = ((A + 2) * r - (A + 3)) * r * r + 1;
+    v[2] = ((A + 2) * (1 - r) - (A + 3)) * (1 - r) * (1 - r) + 1;
+    v[3] = ((A * (2 - r) - 5 * A) * (2 - r) + 8 * A) * (2 - r) - 4 * A;
+    return v;
+}
+#define cubic_coeffs FUNC_CALL(cubic_coeffs)
+
+inline matrix_data_4x4_t FUNC(
+    gather_4x4)(const data_t* data, const size_t n, const size_t c, const long y_topleft, const long x_topleft) {
+    matrix_data_4x4_t s;
+    for (int j = 0; j < 4; ++j)
+        for (int i = 0; i < 4; ++i)
+            s[j * 4 + i] = get_padded(data, n, c, y_topleft + j, x_topleft + i);
```

Comment:
@ozhydkov-lohika  I think it's better to use curly braces in condition statements.

Reply:
resolve